Runtimes/Supplemental/Synchronization/CMakeLists.txt

```diff
@@ -141,6 +141,31 @@ target_link_libraries(swiftSynchronization PRIVATE
   $<$<PLATFORM_ID:Darwin>:swiftDarwin>
   $<$<PLATFORM_ID:Windows>:ClangModules>)
 
+if(WIN32 AND CMAKE_SYSTEM_PROCESSOR STREQUAL i686)
+  # FIXME `-whole-module-optimization` should not be needed. However, without
+  # this, we will not properly optimize. On Windows i686 in particular, we
+  # observe the formation of `__atomic_load` calls rather than the expected
+  # native sequence.
+  #
+  # WMO requires the early-swift-driver to be usable, which is not yet ready on
+  # Windows. Workaround this by creating an empty stub for swiftCore, removing
+  # the import library from the command line, and adding the library search path
+  # associated with it. Due to the autolink, we will still link against the
+  # library from the correct location but will also use an empty file for the
+  # additional input allowing us to perform the WMO.
+  target_compile_options(swiftSynchronization PRIVATE
+    $<$<COMPILE_LANGUAGE:Swift>:-whole-module-optimization>)
```

Comment:
This should check whether `CMP0157` is enabled before setting this. If it is, then it should set [`CMAKE_Swift_COMPILATION_MODE`](https://cmake.org/cmake/help/latest/variable/CMAKE_Swift_COMPILATION_MODE.html) to `wholemodule` instead, or CMake may generate the wrong build graph.